inference: re-architect entry points for more clarity

vtjnash opened the pull request:
The main functional changes are:
 - More nearly separates the optimization steps from the inference loop
   (still some interleaving, due to wanting to cache work results eagerly)
 - Doesn't run `widen_all_const` on the IR before returning to the user
   during reflection (e.g. code-typed). This allows the user to run
   additional passes on it outside of Julia codegen,
   or simply to better see what inference knows.
 - There's now a flag in `Compiler.Params` for whether the global
   inference cache is applicable and should be used (read / write).
   Although this flag is currently not respected, however.

JeffBezanson commented:
Could you undo some of the movement of code between files? It would make it easier to see what changed (and possibly reduce conflicts with my branch).

vtjnash commented:
I had tried to make two commits (code movement then changes), but messed up the intermediate state and had to squash them. I can try to recreate that. But I can probably easily rebase this on top of your branch later. Which branch is it? The changes are primarily in the driver functions (e.g. `typeinf` https://github.com/JuliaLang/julia/pull/27310/files#diff-08648ba14e7c539a4cd73dfa18c2de57R13) and handling the `cache` parameter differently, and then moving all of its helper functions to keep them together.

vtjnash committed:
inference: re-architect entry points for more clarity

The main functional changes are:
 - More nearly separates the optimization steps from the inference loop
   (still some interleaving, due to wanting to cache work results eagerly)
 - Doesn't run `widen_all_const` on the IR before returning to the user
   during reflection (e.g. code-typed). This allows the user to run
   additional passes on it outside of Julia codegen,
   or simply to better see what inference knows.
 - There's now a flag in `Compiler.Params` for whether the global
   inference cache is applicable and should be used (read / write).
   Although this flag is currently not respected, however.
 - Remove some dead code vestiges still left over from the old optimizer.

vtjnash committed:
Compiler: move substitute! function to Meta module

No longer needed by the NewIR optimizer, but useful for some external consumers.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27310#issuecomment-397184932) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/215e439a047d2f82223699725620f5ce2b0ab383/366e262_vs_c8f034c/report.md). cc @ararslan
